[channel-namming] - supplement recommendations (#232)
operator-framework/olm-docs: content/en/docs/best-practices/channel-naming.md

  sourceNamespace: my-operators
```

The [Subscription][subscription] is providing hints to OLM which are used to determine exactly which version of an operator will get deployed onto the cluster, in this example OLM will look for an operator to deploy that belongs to the candidate channel within a specified catalog source.  

Note that exactly which operator version is deployed can depend on more than what you specify in the [Subscription][subscription].  On initial install, OLM will always attempt to install whatever is the head of the specified channel by default.  Settings within the operator’s CSV also are used by OLM to determine exactly which operator version OLM will deploy or upgrade.

## NAMING

> Channel names are used to imply what form of upgrade you want to offer for your operator.  For example, you might have an operator that has a candidate release version which is not supported as well as a version where support is offered.
> 
> Our motivation with this convention is to encourage a better user experience for OLM users to have an intuitive 
> understanding of the level of maturity and supportability of the solutions that will get installed. It can get very complex
> for cluster admins and Operator consumers in general when each solution distribute in the OLM catalog adopts different names
> or different meanings for the same terminologies.

Author: ```suggestion
Channel names are used to imply what form of upgrade you want to offer for your operator. For example, you might have an operator with a candidate release version that is not supported and a version where support is offered.
		

Our motivation with this convention is to encourage a better user experience for OLM users to have an intuitive 
understanding of the level of maturity and supportability of the solutions that will get installed. 
It can get very complex for cluster admins and Operator consumers in general when each solution distributed in the OLM catalog adopts different names or different meanings for the same terminologies.
```

The names you choose are notional and up to you to decide, however, picking good channel names requires some basic guidance.  What is described below are different channel naming conventions that are commonly used by the operator community to denote different operator upgrade use cases.

### Naming Conventions

* Channel names are chosen by operator authors as they see fit to meet their upgrade strategies.
* Channel names are related to the release strategy and not to specific features. ( e.g. we can have an Operator bundle release which is supported and ought to be under the stable channel but with a "tech-preview" and not supported feature. ).

> > Note: If your new operator bundle release contains an API(CRD), which represents new experimental features and not supported (like "tech-preview"), then
> the terminology recommended to being adopted would be in the API version (e.i. group: my.example.com  kind: Backup version: v1alpha1) and **not** in the channel naming. The recommendation for this scenario is to follow up the [Kubernetes API versioning](https://kubernetes.io/docs/reference/using-api/#api-versioning).

Author: ```suggestion
> **Note:** If your new operator bundle release contains an API(CRD), which represents new experimental features and is not supported (like "tech-preview"), then the terminology recommended to adopt would be in the 
API version (e.i. group: my.example.com kind: Backup version: v1alpha1) and **not** in the channel naming. The recommendation for this scenario is to follow up the [Kubernetes API versioning](https://kubernetes.io/docs/reference/using-api/#api-versioning).
```

* Channel names are unique to your operator and do not collide with channel names used by other operator providers.
* Seldom is there a situation where your channel names need to contain information about the Kubernetes or Openshift cluster version they run on.  Only in the case where your operator versions have a dependency on the Kubernetes/Openshift version would you include the cluster version in your channel name.
* You typically would **not** include product names in your channels since the channels are unique to your product and will not collide with other channel names used by other operators.
* You could include or have an operand version in your channel name to advertise to consumers the version of operand they can subscribe to.
* If you do choose to include some version in your channel name, it is important to include an additional identifier, to clarify what that the version number is referring to. A version number could equally be referring to product version (operand version), or operator version - these two don't always match: the operator itself can have different versioning than the product it is managing. 

> * It is recommended to use the major versions of your Operator releases in the channel names at least. So that cluster admins can better plan when they can begin to consume newer versions of your Operators, which contains breaking changes and avoid workflow issues. (e.g. `stable-2.x`)

Reviewer: ```suggestion
* It is recommended to use the major versions of your Operator releases in the channel names at least. So that cluster admins can better plan when they can begin to consume newer versions of your Operators, which contains breaking changes and avoid workflow issues. (e.g. `stable-2`)
```

(if you're going to align w/ the examples below.  Though i have to say, it's not super obvious that "2" is a version number in that name.  @joelanford i think you proposed `stable-2` instead of `stable-2.x`, do you have thoughts here?  maybe it's not worth worrying about, or maybe we recommend `stable-v2`?  I don't know how many products will do a major version only name, vs "stable-2.1" which is less of an issue))

Reviewer 2: To me, its obvious if there are y-named channels there for context `stable-2`, `stable-2.0`, `stable-2.1`, etc. That's what I've typically seen with semver-based docker image versioning (e.g. [nginx](https://hub.docker.com/_/nginx) and [debian](https://hub.docker.com/_/debian)

I'm not opposed to a `v` prefix to make that more obvious though.

Reviewer: > To me, its obvious if there are y-named channels there for context stable-2, stable-2.0, stable-2.1

if, yes.  I don't know how many teams are going to maintain *both* "major-version" channels (foo-X) *and* "minor version" channels(foo-X.Y) though, so i don't know if you can really assume that there will be a "foo-X.Y" that lends context to a "foo-X"

Author: I am OK with we add the prefix`v`.
Indeed I think that is a good idea but then that means adding it for all scenarios.

> * It is recommended to use the major versions of your Operator releases in the channel names at least. So that cluster admins can better plan when they can begin to consume newer versions of your Operators, which contains breaking changes and avoid workflow issues. (e.g. `stable-2.x`)

Author: ```suggestion
* It is recommended to use the major versions of your Operator releases in the channel names at least. So that cluster admins can better plan when they can begin to consume newer versions of your Operators, which contains breaking changes and avoid workflow issues. (e.g. `stable-v2`)
```

  sourceNamespace: my-operators
```

The [Subscription][subscription] is providing hints to OLM which are used to determine exactly which version of an operator will get deployed onto the cluster, in this example OLM will look for an operator to deploy that belongs to the candidate channel within a specified catalog source.  

Note that exactly which operator version is deployed can depend on more than what you specify in the [Subscription][subscription].  On initial install, OLM will always attempt to install whatever is the head of the specified channel by default.  Settings within the operator’s CSV also are used by OLM to determine exactly which operator version OLM will deploy or upgrade.

## NAMING

> Channel names are used to imply what form of upgrade you want to offer for your operator. For example, you might have an operator with a candidate release version that is not supported and a version where support is offered.

Reviewer 3: Do you think it'll be helpful to clarify this point a little more with examples? 
```suggestion
Channel names should imply the nature of official support the user can expect when subscribed to that channel for your operator. For example, the channel name `stable-v2` implies that subscribing to this channel will install an operator with a major version `v2` that you can expect official support for. The channel name `candidate-v2` implies that the user will be installing a candidate release version that does not have official supported.
```

Author: The examples are added, I mean, they exist already in the doc

		

Our motivation with this convention is to encourage a better user experience for OLM users to have an intuitive 
understanding of the level of maturity and supportability of the solutions that will get installed.

> It can get very confusing for cluster admins and Operator consumers in general when each solution distributed in the OLM catalog adopts different names or different meanings for the same terminologies.

Reviewer 3: "It can be very confusing for cluster admins and Operator consumers in general when each solution distributed in the OLM catalog adopts different names or different meaning for the same terminologies" 

Is this sentence really required in this doc? We know it's out motivation to write this entire channel naming convention doc, but just reading the sentence it looks like it has the potential to throw off users with words like "solution", "terminologies" etc. We know what they mean to us who's been looking at these "solutions" and "terminoliges" for a long time now but a new operator author will probably be very confused with this sentence (and thereby increase the cognitive overload needed for reading this doc)

Author: One of the goals is to clarify why we need the convention.
Besides, it shows obvious for us that have not been obvious to everybody.

The names you choose are notional and up to you to decide, however, picking good channel names requires some basic guidance.  What is described below are different channel naming conventions that are commonly used by the operator community to denote different operator upgrade use cases.

### Naming Conventions

* Channel names are chosen by operator authors as they see fit to meet their upgrade strategies.

> * Channel names are related to the release strategy and not to specific features. ( e.g. we can have an Operator bundle release which is supported and ought to be under the stable channel but with a "tech-preview" and not supported feature. ).

Reviewer 3: ```suggestion
* Channel names should communicate the release strategy, official support level and the level of maturity, but **not** specific features. ( e.g. a channel with Operator versions which has official support ought to be have the word `stable` in the channel name, whereas a channel with Operator versions with a non-supported feature should have the phrase `tech-preview`in it's name. ).
```

> > **Note:** If your new operator bundle release contains an API(CRD), which represents new experimental features and

Reviewer 3: ```suggestion
> **Note:** If your new operator version release contains an API(CRD), which represents new experimental features and
```

#### Example 2

By following the recommendation, you would provide the major versions of the Operators
in the channel naming. So that your consumers can be subscribed to a channel where
they can safely install the latest versions published on it and avoid workflow
issues on their cluster because they know that none of the solutions which will be
distributed within contains breaking changes:

| Channels for Operator version(s) 2.x | 
| :------------- | 
| candidate-2.x |
| fast-2.x | 

> | stable-2.x |

Reviewer 2: I think I'd leave off the `.x` suffix to align with major/minor channel names and make it more clear that these channels track all releases within a major version.

#### Example 2

By following the recommendation, you would provide the major versions of the Operators
in the channel naming. So that your consumers can be subscribed to a channel where
they can safely install the latest versions published on it and avoid workflow
issues on their cluster because they know that none of the solutions which will be
distributed within contains breaking changes:

> | Channels for Operator version(s) 2.z | 

Reviewer: ```suggestion
| Channels for Operator version(s) 2.x | 
```

#### Example 2

By following the recommendation, you would provide the major versions of the Operators
in the channel naming. So that your consumers can be subscribed to a channel where
they can safely install the latest versions published on it and avoid workflow
issues on their cluster because they know that none of the solutions which will be
distributed within contains breaking changes:

| Channels for Operator version(s) v2.x | 
| :------------- | 
| candidate-v2 |
| fast-v2 | 
| stable-v2 |

> **Attention:** The versions on the channel names are related to the Operator version and not its Operand.

Reviewer 4: Nit: I would add "s" to "Operand"

#### Example 3 (Recommended Option is the most common scenarios)

An operator wants to be supported at various operator major/minor versions.  
For example, you might have an operator version at `1.3` and also at `2.4` that 
you need or want to offer support for at the same time.  However, you might 
not want to have OLM upgrade users to the `2.4` Operator but instead keep them 
upgrading within the `1.3` versions. In that case, you would end up with channels as 
recommended above but with major/minor version information applied as follows:

| Channels for v1.3 | Channels for v2.4 |
| :------------- | :----------- |
| candidate-v1.3 | candidate-v2.4 |
| fast-v1.3 | fast-v2.4 |
| stable-v1.3 | stable-v2.4 | 

> **Attention:** The versions on the channel names are related to the Operator version and not its Operand.

Reviewer 4: Nit: I would add "s" to "Operand"

#### Example 2

By following the recommendation, you would provide the major versions of the Operators
in the channel naming. Consumers can then subscribe to a channel with confidence that none of the versions in the 
channel will introduce breaking changes:

| Channels for Operator version(s) v2.x | 
| :------------- | 
| candidate-v2 |
| fast-v2 | 
| stable-v2 |

**Attention:** The versions on the channel names are related to the Operator version and not its Operands.

#### Example 3 (Recommended Option is the most common scenarios)

> An operator wants to be supported at various operator major/minor versions.  

Reviewer 3: ```suggestion
An operator should be supported at various operator major/minor versions.  
```

> For example, you might have an operator version at `1.3` and also at `2.4` that 

Reviewer 3: ```suggestion
For example, you might have a `v1.3` and also a `v2.4` releases of your operator that 
```

you need or want to offer support for at the same time.  However, you might 

> not want to have OLM upgrade users to the `2.4` Operator but instead keep them 

Reviewer 3: ```suggestion
not want to have OLM upgrade users to the `v2.4` Operator but instead keep them 
```

Channel promotion is the notion of moving an operator from one channel to 
another.  For example, consider the case where you have an operator version 
`1.0.1` which is found in a candidate channel, then you might decide to offer 
support for that version and want to move it to a stable channel.  

Today, channel promotion is achieved by creating a new operator version 
(`1.0.2`) that is labeled with the channel(s) you want to promote to (as well 
as any channels you want to keep it in).

# UPGRADES 

OLM provides a variety of ways to specify updates between operator versions. Before
continue in the following recommendations, make sure you understand its options by checking [Creating an update graph with OLM][upgrade-graph].

## Recommended upgrade path for patch release (bug fixes)

Within a channel each patch release should be directly upgradeable. Use skip or skipRange
to ensure its behaviour. 

#### Example

If you have the channel `stable-3.7` where the head of channel is `3.7.10` and then, you will provide a new patch release
with a bug fix using the Operator bundle version `3.7.11` then, ensure that you configure your new publication to skip all 

> that was published in under the channel `stable-3.7` (e.g. skipRange: `">= 3.7.0 < 3.7.11"` replaces: `3.7.10` ).  

Reviewer 2: +1 Ben

The problem is that most readers are likely still using the sqlite imperative workflow, so its very difficult (maybe impossible) to track a moving target from the previous y stream.

This would be very easy to implement in FBC, especially with a veneer that implements all the logic that's described.

So that, you can ensure that your users can move forward easier to the latest version since in this range `">= 3.7.0 < 3.7.11"`
you just provided bug fixes and not new features or breaking changes.

## Recommended upgrade path cross-channel

After to be subscribed to a channel, OLM users are able to update their [Subscription][subscription] to
consume the solutions from another channel, for example, by moving from the fast channel to
the stable one.

Because of this, it is recommended configure your bundles versions to allow cross-channel updates at least 
between two consecutive major/minor versions (e.g. by having installed the head of the channel `stable-2.9` 
your users can get installed the first version under `stable-2.10`)

#### Example

If you have published solutions on the channel `stable-3.5`

> then, you would like to begin to distribute Operator bundle versions `3.6.x`

Reviewer: ```suggestion
then, you would like to begin to distribute Operator bundle versions `3.6.z`
```

(versioning is X.Y.Z)

that would mean distributed a new bundle under `stable-3.6` skipping
all versions which are provided via the `stable-3.5` channel. 

So that, you can ensure that your users can move forward from any version that they have installed promoted via the 

> `stable-3.5` channel to the first version of the channel `stable-3.6`. (e.g. `skipRange: ">=3.5.0 <3.6.0"`)

Reviewer: Caveat to call out here is this means that the operator author must be comfortable that any 3.6.z version can upgrade from any 3.5.z version.

instead of `3.5.0` is a mechanism to constrain the support matrix when necessary, it's not the default recommendation.

**Note** If you would like to ensure that the users of your Operator still able to install the Operator bundle version
`3.6.29` by using the option `startingCSV` added manually in the subscription then, you will need to also use the `replaces: 3.6.29`
to ensure the older Operator bundle does not get pruned from the index catalog via the OLM resolver.

#### Attention (Be aware of the following example scenario)

If you have the channel `stable-v3.7` where the head of channel is `3.7.10` and then, you provide a new patch release
with a bug fix using the Operator bundle version `3.7.11` then, imagine that you configure your new publication to skip all
that was published under the channel `stable-v3.7` (e.g. skipRange: `">= 3.7.0 < 3.7.11"` replaces: `3.6.z <latest release on 3.6 channel>` ).

You could ensure that your users can move forward easier to the latest version since in this range `">= 3.7.0 < 3.7.11"`
as they will be able to install the new `3.7.11` release from  `3.6.z <latest release on 3.6 channel>` directly 

> instead of get installed first the Operator version `3.7.0` which we know that have bug fixes in place already.

Reviewer: ```suggestion
instead of having to upgrade through Operator version `3.7.0` which may contain bugs that are already fixed in both the 3.6.z version they had installed, and the 3.7.11 version they are trying to get to.
```

> However, **attention** that means you would no longer able to provide any patch release to the minor channel `stable-v3.6`.    

Reviewer: ```suggestion
However, this approach has negative implications when you provide your next patch release to the minor channel `stable-v3.6`.    
```

Note that if you publish `3.6.latest+1` when this version comes out, and your users upgrades to it, they will have no way to

> upgrade from `3.6.latest+1` to any solution published under the channel `stable-v3.7`.

Reviewer: ```suggestion
upgrade from `3.6.latest+1` to any solution published under the channel `stable-v3.7`, until you publish a new 3.7.z version that replaces 3.6.latest+1.
```

Channel promotion is the notion of moving an operator from one channel to 
another.  For example, consider the case where you have an operator version 
`1.0.1` which is found in a candidate channel, then you might decide to offer 
support for that version and want to move it to a stable channel.  

Today, channel promotion is achieved by creating a new operator version 
(`1.0.2`) that is labeled with the channel(s) you want to promote to (as well 
as any channels you want to keep it in).

# UPGRADES 

OLM provides a variety of ways to specify updates between operator versions. Before
continue in the following recommendations, make sure you understand its options by checking [Creating an update graph with OLM][upgrade-graph].

## Recommended upgrade path for patch release (bug fixes)

Within a channel each patch release should be directly upgradable to the `head channel` from the latest version shipped 

> in the minor channel behind. Use skips or skipRange to provide this behaviour. (i.e. using the skipRange in the 

Reviewer: ```suggestion
in the previous minor version channel. Use skips or skipRange to provide this behaviour. (i.e. using the skipRange in the 
```

> `stable-3.6` patch for `3.5.<latest>: >= 3.5.<first z of the minor channel>. < 3.6.<new pacth release>`)

Reviewer: ```suggestion
`stable-3.6` patch for `3.5.<latest>: >= 3.5.<first z of the minor channel>. < 3.6.<new patch release>`)
```

aside from the typo, i can't understand what this is saying.

should it be this instead:
`stable-3.6` patch for `3.6.<latest>` would mean setting the skipRange to be `>= 3.5.<first z of the minor channel>. < 3.6.<new patch release>`, where 3.5.z represents the oldest 3.5.z version for which you can provide direct upgrade support to 3.6.latest)

#### Example

Let's imagine that you will publish the Operator bundle `3.6.30` (patch release under the channel `stable-3.6`) and
that you have published so far:

- Operator bundles versions from `3.5.0` to `3.5.25` under the `stable-3.5` channel
- Operator bundles versions from `3.6.0` to `3.6.29` under the `stable-3.6` channel

Then, in this case, your newer patch release `3.6.30` would be configured with `skipRange: >=3.5.25 < 3.6.30` in

> order to only supports upgrading to the newest `3.6.z` from a pretty recent `3.5.z`

Reviewer: They shouldn't necessarily use 3.5.25, they should use the oldest 3.5.z version that they can support upgrades from.

they can use 3.5.25 if, for some reason, they don't think upgrades from something older like 3.5.24 (or more likely, 3.5.0) to 3.6.latest will work.  But it the ideal situation is that they do allow upgrades from 3.5.0 to 3.6.latest (i.e. `skipRange: >=3.5.0 < 3.6.30`).  The ability to set it to 3.5.25 (or 3.5.17, or something else) instead of 3.5.0 is a mechanism to constrain the support matrix *when necessary*, it's not the default recommendation.

**Note** If you would like to ensure that the users of your Operator still able to install the Operator bundle version
`3.6.29` by using the option `startingCSV` added manually in the subscription then, you will need to also use the `replaces: 3.6.29`

> for this Operator bundle does not get pruned from the index catalog via the OLM resolver.

Reviewer: ```suggestion
to ensure the older Operator bundle does not get pruned from the index catalog via the OLM resolver.
```

#### Attention (Be aware of the following example scenario)

If you have the channel `stable-3.7` where the head of channel is `3.7.10` and then, you provide a new patch release
with a bug fix using the Operator bundle version `3.7.11` then, imagine that you configure your new publication to skip all
that was published under the channel `stable-3.7` (e.g. skipRange: `">= 3.7.0 < 3.7.11"` replaces: `3.6.z <latest release on 3.6 channel>` ).

You could ensure that your users can move forward easier to the latest version since in this range `">= 3.7.0 < 3.7.11"`
as they will be able to install the new `3.7.11` release from  `3.6.z <latest release on 3.6 channel>` directly 
instead of get installed first the Operator version `3.7.0` which we know that have bug fixes in place already.

However, **attention** that means you would no longer able to provide any patch release to the minor channel `stable-3.6`.    
Note that if you publish `3.6.latest+1` when this version comes out, and your users upgrades to it, they will have no way to

> upgrade from `3.6.latest+1` to any solution published under the channel `stable-3.7`.

Reviewer: suggest this instead for the above paragraph:

However, that means that if/when you publish a `3.6.latest+1` and a user upgrades to it, they will not
have a path to `3.7` because `3.6.latest+1` is not included in the skipRange defined by any existing `3.7.z` bundle.  You would have to then publish a new `3.7.z` bundle that expands the skipRange to include 3.6.latest+1 before users would be able to upgrade to a 3.7.z version.

Reviewer: (this is a restriction that becomes less problematic with FBC when operator authors are able to update the skipRange of a previously published bundle, so they could extend the skipRange of 3.7.11 to include 3.6.latest+1, but of course we aren't there yet)

Author: I think a follow up here will be to change this doc for FBC or tweak it. 
IMO I do not think that we should only FBC instructions any way since it will be optional in the future for the authors and many of them might never adopt it.

Reviewer: >  I do not think that we should only FBC instructions any way since it will be optional in the future for the authors and many of them might never adopt it.

at a minimum i agree it's probably too early to do so.

Reviewer 2: this is reminiscent of the discussion we had many months ago when we replaced some imperative `opm registry add` docs with the new FBC docs.

the main branch of the upstream docs site focuses on FBC, not the deprecated sqlite workflows.

Reviewer 2: to reframe this as a question, why should we recommend that upstream users keep using deprecated workflows when FBC exists and helps solve a bunch of these problems?

if there are pipelines that have not adopted FBC, those pipelines likely need to have documentation about their requirements and assumptions.

IMO the upstream OLM docs site should help catalog maintainers and pipeline builders understand how to get off the deprecated path and onto FBC.

Author: Closing this one since we spoke shows that we agree that its goal is only to add what is missing we probably want to work in the future to see how we can shape it for FBC.

## Recommended upgrade path cross-channel

After subscribing to a channel, OLM users are able to update their [Subscription][subscription] to
consume the solutions from another channel, for example, by moving from the fast channel to
the stable one.

Because of this, it is recommended configure your bundles versions to allow cross-channel updates at least 
between two consecutive major/minor versions (e.g. by having installed the head of the channel `stable-2.9` 

> your users can get installed the first version under `stable-2.10`)

Reviewer: this whole section seems redundant to the previous section(` Recommended upgrade path for patch release (bug fixes)`) and the recommendation is worse (because it recommends upgrade paths which take the user through the X.Y.0 version of a minor stream, instead of directly to X.Y.latest).

everything recommended in the previous section already allows for a user to move from "stable-2.9" to "stable-2.10" if it is done correctly.  (i.e. the previous section covers both patch upgrades *and* minor version upgrades).

I suggest dropping this section and simply retitling the previous section to clarify that it's upgrade graph structure recommendations generally.

Channel promotion is the notion of moving an operator from one channel to 
another.  For example, consider the case where you have an operator version 
`1.0.1` which is found in a candidate channel, then you might decide to offer 
support for that version and want to move it to a stable channel.  

Today, channel promotion is achieved by creating a new operator version 
(`1.0.2`) that is labeled with the channel(s) you want to promote to (as well 
as any channels you want to keep it in).

# UPGRADES 

OLM provides a variety of ways to specify updates between operator versions. Before

> continue in the following recommendations, make sure you understand its options by checking [Creating an update graph with OLM][upgrade-graph].

Reviewer 4: Proposing "Before we continue with the following recommendations, make sure you understand the options by checking [Creating an update graph with OLM][upgrade-graph]."

## Recommended upgrade path

Within a channel each patch release should be directly upgradable to the [HEAD of channel](/docs/glossary/#channel-head). 

> Use skips or skipRange to provide this behaviour. (i.e. if you followed the above `Example 3` that means 

Reviewer 3: We should probably link to the `skips` and `skipsRange` sections of the doc site from here

Author: Usually, I would agree with. But then it can bring confusion Why?
skip/skipRange/replaces configured via CSV still supported and valid indeed with FBC format. 
However, our latest version of this doc ONLY explains how to do it with the new FBC options and we remove completed this explanation about how to do it using the CSV.

use skipRange to publish a patch for `3.6.z`under the`stable-v3.6` would mean setting the skipRange to be 
`>= 3.5.z < 3.6.z`, where `3.5.z` represents the oldest `3.5.z` version for which 
you can provide direct upgrade support to `3.6.z` latest)

#### Example

Let's imagine that you will publish the Operator bundle `3.6.30` (patch release under the channel `stable-v3.6`) and
that you have published so far:

> - Operator bundles versions from `3.5.0` to `3.5.25` under the `stable-v3.5` channel

Reviewer 3: ```suggestion
- Operator bundles versions from `v3.5.0` to `v3.5.25` under the `stable-v3.5` channel
```

This is a nit, for all the versions quoted in the rest of the section too, just to make it look a little more clean by following the semver convention

- Operator bundles versions from `3.6.0` to `3.6.29` under the `stable-v3.6` channel

Then, in this case, your newer patch release `3.6.30` would be configured with `skipRange: >=3.5.25 < 3.6.30` in
order to only supports upgrading to the newest `3.6.z` from a pretty recent `3.5.z`. 

You do not need necessarily to use `3.5.25`. Therefore, you should 
use the oldest `3.5.z` version that supports upgrades from. Be aware that the ability to configure the path 
using this example from an Operator version like `3.5.25` _(or `3.5.17`, or something else)_ 
instead of `3.5.0` is a mechanism to constrain the support matrix when necessary, it's not the default recommendation.

> **Note** If you would like to ensure that the users of your Operator still able to install the Operator bundle version

Reviewer 4: Wording; "that the users of your Operator still able" -> "that the users of your Operator are still able"

`3.6.29` by using the option `startingCSV` added manually in the subscription then, you will need to also use the `replaces: 3.6.29`
to ensure the older Operator bundle does not get pruned from the index catalog via the OLM resolver.

> #### Attention (Be aware of the following example scenario)

Reviewer 4: Proposing: "Be aware of the following scenario"

> If you have the channel `stable-v3.7` where the head of channel is `3.7.10` and then, you provide a new patch release

Reviewer 3: ```suggestion
If you have the channel `stable-v3.7` where the head of channel is `v3.7.10`, and you have a new patch release
```

> with a bug fix using the Operator bundle version `3.7.11` then, imagine that you configure your new publication to skip all

Reviewer 3: ```suggestion
with a bug fix using the Operator bundle version `v3.7.11` , and you configure `v3.7.11` to skip all
```

> that was published under the channel `stable-v3.7` (e.g. skipRange: `">= 3.7.0 < 3.7.11"` replaces: `3.6.z <latest release on 3.6 channel>` ).

Reviewer 3: ```suggestion
operator bundles published in the channel `stable-v3.7` (e.g.`v3.7.11` mentions  skipRange: `">= 3.7.0 < 3.7.11"` replaces: `3.6.z <latest release on 3.6 channel>` ).
```

> By doing so you could ensure that your users can more easily upgrade to the latest version since they will 

Reviewer 3: ```suggestion
, you could ensure that your users can more easily upgrade to the latest version since they will 
```

as they will be able to install the new `3.7.11` release from  `3.6.z <latest release on 3.6 channel>` directly
instead of having to upgrade through Operator version `3.7.0` which may contain bugs that are already fixed in both the `3.6.z` version 
they had installed, and the `3.7.11` version they are trying to get to.

> 

Reviewer 4: Proposing: "By doing so you could ensure that your users can more easily upgrade to the latest version since they will be able to install the new `3.7.11` release directly from `3.6.z <latest release on 3.6 channel>` instead of having to upgrade through Operator version `3.7.0` which may contain bugs that are already fixed in both the `3.6.z` version they had installed, and the `3.7.11` version they are moving to."

However, this approach has negative implications when you provide your next patch release to the minor channel `stable-v3.6`.  

> Note that if you publish `3.6.latest+1` when this version comes out, and your users upgrades to it, they will have no way to

Reviewer 4: Nit: "and your users upgrade to it"
